Support parsing non-nul-terminated PEM strings

Koromix opened the pull request:
## Description

Direct or indirect use of `mbedtls_pem_read_buffer()` is confusing because the string must be nul-terminated.

This poses a few problems:
- Other libraries (e.g. curl) often need to make a copy of their own buffer to add a terminating NUL byte before calling into Mbed TLS.
- Several Mbed TLS functions (e.g. mbedtls_x509_crt_parse) seem to support non-nul-terminated buffers but this fails with PEM data.

Fix this situation by introducing a new function, `mbedtls_pem_read_buffer_with_len()`. The signature of `mbedtls_pem_read_buffer()` is left unchanged, and the function just forwards to mbedtls_pem_read_buffer_with_len() (with strlen).

A new utiltiy function is added, `mbedtls_platform_strnstr()` to replace the use of strstr. The signature matches strnstr(): https://www.freebsd.org/cgi/man.cgi?query=strnstr&sektion=3

## Status
Ready

## Requires Backporting
No

## Migrations
Not in this version: a new API is introduced and the existing API is left unchanged.

## Todos
- Determine if mbedtls_platform_strnstr() is appropriate and in the correct place (platform_util)?
- Rename mbedtls_pem_read_buffer_with_len() or even replace mbedtls_pem_read_buffer()?

daverodgman commented on library/x509_crl.c:
```diff
@@ -546,15 +546,10 @@ int mbedtls_x509_crl_parse( mbedtls_x509_crl *chain, const unsigned char *buf, s
     {
```
The docs in the declarations of `mbedtls_x509_crl_parse_der` and `mbedtls_x509_crl_parse` in `include/mbedtls/x509_crl.h` need updating - `buflen` no longer needs to allow for a terminating null byte.

Same comment applies to the crt and csr functions.

daverodgman commented on library/dhm.c:
```diff
@@ -524,14 +524,10 @@ int mbedtls_dhm_parse_dhm( mbedtls_dhm_context *dhm, const unsigned char *dhmin,
 #if defined(MBEDTLS_PEM_PARSE_C)
     mbedtls_pem_init( &pem );
 
```
Need to update the docs in dhm.h, which refer to buflen needing to allow for a trailing NULL.

Koromix replied:
Same

daverodgman reviewed (changes requested):
I think we also need to update the documentation for the pk functions in pk.h, i.e. `mbedtls_pk_parse_key` and `mbedtls_pk_parse_public_key`, as per @gilles-peskine-arm comments in #3896

Koromix commented on library/pem.c:
```diff
@@ -225,6 +225,20 @@ static int pem_aes_decrypt( unsigned char aes_iv[16], unsigned int keylen,
                      const unsigned char *data, const unsigned char *pwd,
                      size_t pwdlen, size_t *use_len )
+{
+    int ret;
```
Done in the latest pushed commits.

Koromix commented:
Thanks for the review :)

I've rebased my commits on development, and made the following changes:

* Added tests for mbedtls_platform_strnstr in tests/suites/test_platform.function (and .data), which is a new file I made for this function.
* Added tests for mbedtls_pem_read_buffer_with_len in tests/suites/test_pem.function (and .data), including one with a valid PEM non-NUL terminated buffer (using data_t/unhexify).
* Changed the documentation of the various callers of mbedtls_pem_read_buffer_with_len, I simply removed the notes about the need to NUL-terminate PEM data.

gilles-peskine-arm commented on tests/suites/test_suite_platform.function:
```diff
@@ -0,0 +1,23 @@
+/* BEGIN_HEADER */
+
+/* This test module exercises the timing module. Since, depending on the
+ * underlying operating system, the timing routines are not always reliable,
+ * this suite only performs very basic sanity checks of the timing API.
+ */
```
Copypasta

gilles-peskine-arm commented on tests/suites/test_suite_pem.function:
```diff
@@ -51,3 +51,26 @@ exit:
     mbedtls_pem_free( &ctx );
 }
 /* END_CASE */
+
```
Rather than duplicate the existing test data and much of the test code, please call `mbedtls_pem_read_buffer_with_len` from `mbedtls_pem_read_buffer` as well, using a copy of the buffer without a null terminator (`ASSERT_ALLOC( buf, strlen( data ) )`). And call `mbedtls_pem_read_buffer_with_len` on `data` with the null terminator as well, to check that this also works. And check that the value returned for `use_len` is the same in all cases.

gilles-peskine-arm replied:
Note that https://github.com/ARMmbed/mbedtls/pull/5508 and https://github.com/ARMmbed/mbedtls/pull/5504 are also touching PEM testing (seems to be a hot topic at the moment!).

gilles-peskine-arm reviewed (changes requested):
The library change looks good to me but the tests and documentation need work.

gilles-peskine-arm commented:
#5504 has been merged, so please rebase on top of `development`.

gilles-peskine-arm commented on library/pem.c:
```diff
@@ -239,16 +274,17 @@ int mbedtls_pem_read_buffer( mbedtls_pem_context *ctx, const char *header, const
 
-    s2 = (unsigned char *) strstr( (const char *) data, footer );
+    s2 = (unsigned char *) mbedtls_pem_strnstr( (const char *) data, footer, datalen );
 
```
Now the code that skips whitespace after the header and footer can overflow the input buffer. For the header, it can only happen in pathological cases where the footer consists only of whitespace, but that's not a documented exclusion. For the footer, it's quite likely, and in fact the CI found it with `tests/data_files/cli-rsa.key`.

Please fix this and add unit tests: `pk_parse_key` on `tests/data_files/cli-rsa.key`, and some more general test cases on `mbedtls_pem_read_buf_with_len` directly.

(That's what's causing most of the CI failures.)

gilles-peskine-arm reviewed (changes requested):
> I haven't yet updated the DHM/PK/X509 tests, but it should be noted that the new code paths are partly tested through the simplification of load_file (in dhm.c) and mbedtls_pk_load_file (in pkparse.c), which don't append a nul byte.

I'm ok with those as test cases for the no-null-byte case, but we do need to make sure we also have a positive test case with a trailing null byte.

> I don't think we should try to use the paltform's implementation of strnstr if it exists, for simplicity. But I'm ready to do this if you think it would be better.

`strnstr` is not very common, so I agree, let's keep this simple. We can make it a platform function later if someone actually needs it.

Koromix commented:
Thanks for the review, and sorry for the delay, busy week at work. I'll get back to you in a few days :)

daverodgman commented on tests/suites/test_suite_pem.function:
```diff
@@ -32,30 +33,82 @@ exit:
+
+        mbedtls_pem_free( &ctx );
+        ret = mbedtls_pem_read_buffer_with_len( &ctx, "^^^", "$$$", copy, (size_t)copy_end,
+                    (unsigned char *)valid_pwd, strlen( valid_pwd ), &use_len );
```
Should also validate the returned value in use_len (i.e. that it's consistent with the previous calls to mbedtls_pem_read_buffer_with_len and mbedtls_pem_read_buffer).

Koromix replied:
Not sure what to do here, it's not supposed to be consistent because we don't use the same header and footer, at least I think? This call always fails with MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT, so use_len is unchanged.

daverodgman replied:
Agreed, not sure what my original comment was suggesting.

gilles-peskine-arm commented:
There's a merge conflict and [at least one review comment that hasn't been addressed](https://github.com/Mbed-TLS/mbedtls/pull/5411#pullrequestreview-905980960) (I haven't checked earlier comments). Can you finish the rework, and rebase on top of the `development` branch? (Please keep the same history when you rebase, to facilitate reviewing with `git range-diff`.) If not we'll try to find someone else to finish this.

Koromix commented:
Sorry, kind of forgot about this... I've rebased the code on top of development (without squashing anything) and made some minor changes in the test code.

Koromix commented:
And I've changed the tests to work on MinGW (mbedtls_snprintf does not seem to support %n in MinGW builds).

tom-cosgrove-arm commented:
Apologies for the long delay in reviewing here.

@davidhorstmann-arm has just updated for code style - @Koromix can you fix the conflicts then we will re-review?

Koromix commented:
Hi!

I've just done so :)
I had to change `tests/suites/test_suite_pkwrite.function` a bit to fix pkwrite tests failures.

tom-cosgrove-arm commented:
Hmm, the Travis error was

```
test_suite_dhm .................................................... dhm.c:525:13: runtime error: applying zero offset to null pointer
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior dhm.c:525:13 in 
FAIL
```

- is that due to the change in this PR?

Anyway, @Koromix would you mind fixing the conflicts please?
